### PR TITLE
Fix MODULES-1919: "bind" statements should be sorted by IP, not lexigraphically

### PR DESCRIPTION
**Pull request by TomOnTime**
Fix MODULES-1919

### CODE REVIEW

**Comment by antaflos** [comment]
How does this behave when the `bind` parameter contains something other than IP addresses? See for example https://github.com/puppetlabs/puppetlabs-haproxy/blob/master/spec/defines/frontend_spec.rb#L196-L205

Also, why this change? Does order really matter when specifying `bind` options?

File: templates/fragments/_bind.erb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <% require 'ipaddr' -%>
 <%- if @bind -%>
-<%- @bind.sort.map do |address_port, bind_params| -%>
+<%- @bind.sort_by { |address_port, bind_params| address_port.split('.').map{ |octet| octet.to_i} }.map do |address_port, bind_params| -%>
```
Comment (antaflos) [inline templates/fragments/_bind.erb]: Perhaps this sort code should first partition the keys into two arrays, one containing only the things that look like IP addresses and one containing everything else, then sort both, concatenate them again and process as before.

And what about IPv6 addresses?

**Comment by tlimoncelli** [comment]
Why this change?
It is not just cosmetic. While debugging haproxy.cfg files we found the lexicographical sorting was making it difficult to spot errors that would otherwise be obvious.

How does this behave with non IPs?
It sorts them just fine, alphabetically. There are spec tests that show it works as one would hope.

Partitioning the list into things that look like IPs and others:
Luckily thanks to the designers of ASCII, the non-IPs sort to the top of the list. Therefore, the results are approximately the same without the additional coding.  For example, the test case you pointed to formats as:
```
listen apache
  bind fd@${FD_APP1} 
  bind :443,:8443 ssl crt public.puppetlabs.com no-sslv3
  bind /var/run/ssl-frontend.sock user root mode 600 accept-proxy
  bind 1.1.1.1:80 
  bind 2.2.2.2:8000-8010 ssl crt public.puppetlabs.com
  balance roundrobin
  option tcplog
  option ssl-hello-chk
```

What about IPv6?
Right now IPv6 will sort lexicographically since it doesn't include any periods in the address.   I'd like to improve this in the future. I think this is beyond my ruby skills and I'd appreciate help.  IPv6 support in this module is pretty lacking already.

**Comment by antaflos** [comment]
I think that sounds good and is reasonable to implement, yes. So :+1: from me, whatever that's worth :)

I am wondering about the failing tests, though. Seems the main problem is Ruby 1.8.7, but https://travis-ci.org/puppetlabs/puppetlabs-haproxy/jobs/57502537 and https://travis-ci.org/puppetlabs/puppetlabs-haproxy/jobs/57502538 fail with a `NameError` and `uninitialized constant RSpec::Support::Differ::Encoding` while in https://travis-ci.org/puppetlabs/puppetlabs-haproxy/jobs/57502531 Ruby seems to think `:443 ...` should come before `fd@${FD_APP1}` (https://travis-ci.org/puppetlabs/puppetlabs-haproxy/jobs/57502531#L414). No idea what to make of this.

**Comment by igalic** [comment]
ping @hunner for some… enlightenment…

**Comment by antaflos** [comment]
This was discussed briefly on the puppet-users mailing list and answered by the great jcbollinger: https://groups.google.com/forum/#!topic/puppet-users/9xPXkGW6Pk8

Quoting:
> but it looks like the sort order you are trying to apply does not very effectively cover the space of keys it may be used to sort.  In particular, what order is supposed to be used when the bind_address does not start with a dotted-quad IP address?  Some order will fall out from your code, but there's plenty of room for implementation-defined behavior there.

He also provides an implementation for the `sort_by` block that looks good, but it requires `ipaddr`.

**Comment by TomOnTime** [comment]
Thanks to @jcbollinger's better algorithm, it now passes all tests.  PTAL

**Comment by tphoney** [comment]
@TomOnTime we will review this again in the thursday community meeting

**Pull request closed without merging**